src/operator/tensor/diag_op-inl.h

```diff
@@ -104,42 +143,132 @@ inline bool DiagOpType(const nnvm::NodeAttrs& attrs,
   return (*out_attrs)[0] != -1;
 }
 
-template<int req>
+template<int ndim, int req, bool back>
```

Comment:
should ndim here be index_t too? or maybe im going to far with the index_t.... ;-)

Reply:
Isn't that way too large for the number of dimensions? :)

Reply:
what about for 1x1x1x1x1x1x1xx1x1x1x.......x1x1x1x1x1x1 tensors?

Reply:
`ndim` here can only be 2 or 3 because I do some merge first.